packages/injectable/ogre-tools-injectable.d.ts

```diff
@@ -1,160 +1,144 @@
 
-  type InjectionToken<TInstance, TInstantiationParameter> = {
-    template: TInstance;
-    instantiationParameter: TInstantiationParameter;
+  export interface InjectionToken<
+    Instance,
+    InstantiationParam = void,
+  > {
+    template: Instance;
+    instantiationParameter: InstantiationParam;
     key: Symbol;
-  };
+  }
+
+  export type OptionalParam<InstantiationParam> = InstantiationParam extends void
+    ? []
+    : [InstantiationParam];
 
-  interface CommonInjectable {
+  export type Injectable<
+    Instance extends InjectionInstance,
+    InjectionInstance,
+    InstantiationParam,
+  > = {
     id: string;
     setup?: (di: DiContainerForSetup) => void | Promise<void>;
     causesSideEffects?: boolean;
-    lifecycle?: ILifecycle;
-  }
-
-  interface InjectionTokenInjectable<
-    TInjectionToken extends InjectionToken<unknown, unknown>,
-  > extends CommonInjectable {
-    injectionToken: TInjectionToken;
-
-    instantiate: (
-      di: DiContainerForInstantiate,
-      instantiationParameter: InferFromToken<TInjectionToken>[1],
-    ) => InferFromToken<TInjectionToken>[0];
-  }
-
-  interface NormalInjectable<TInstance, TInstantiationParameter>
-    extends CommonInjectable {
-    instantiate: (
-      di: DiContainerForInstantiate,
-      instantiationParameter: TInstantiationParameter,
-    ) => TInstance;
-  }
+    injectionToken?: InjectionToken<InjectionInstance, InstantiationParam>;
+    instantiate: Instantiate<Instance, InstantiationParam>;
+  } & (
+    InstantiationParam extends void
+      ? ({
+        lifecycle?: typeof lifecycleEnum.singleton;
+      } | {
+        lifecycle: typeof lifecycleEnum.transient | KeyedSingleton;
+      })
+      : {
+        lifecycle: typeof lifecycleEnum.transient | KeyedSingleton;
+      }
+  );
 
   export function getInjectable<
-    TInjectionToken extends InjectionToken<unknown, unknown>,
+    Instance extends InjectionInstance,
+    InjectionInstance,
+    InstantiationParam = void
   >(
-    options: InjectionTokenInjectable<TInjectionToken>,
-  ): InjectionTokenInjectable<TInjectionToken>;
+    options: Injectable<Instance, InjectionInstance, InstantiationParam>,
+  ): typeof options;
 
-  export function getInjectable<TInstance, TInstantiationParameter>(
-    options: NormalInjectable<TInstance, TInstantiationParameter>,
-  ): NormalInjectable<TInstance, TInstantiationParameter>;
-
-  export function getInjectionToken<TInstance, TInstantiationParameter = void>({
+  export function getInjectionToken<
+    Instance,
+    InstantiationParam = void,
+  >({
     id: string,
-  }): InjectionToken<TInstance, TInstantiationParameter>;
+  }): InjectionToken<Instance, InstantiationParam>;
+
+  type AsyncReturnable<IsAsync extends boolean, Instance> = IsAsync extends true
+    ? Promise<Instance>
+    : Instance;
+
+  export interface Inject<IsAsync extends boolean> {
+    <Instance>(
+      key: Injectable<Instance, any, void> | InjectionToken<Instance, void>,
```

Comment:
Prefer `unknown` over `any` here, as it is more deliberate, and semantically correct.